Quiz foundations & data connection

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

- Applies several changes to the root quiz creation page to closer match the Figma specs
- Connects the `useQuizCreation` module to the UI in many places (noted below)
- User can Add and remove sections
- User can select a new active section (as reflected by that section's ID being shown on the page temporarily)
- User can open the "edit" side panel
- User can open the "select resources" side panel (by way of the "+ Add Questions" button)
- The user has an intuitive flow through the side panels. The user has two paths to the "Select resources" panel. They can click "Add Questions" on the root page -- or they can click the "Select resources" button in the "Edit" panel. If they get there from the root page, they see an X to close the side panel. If they get there via the "Edit" panel, they see a back arrow to go back to the previous panel.

**To be done in follow-up issues**

- [x] https://github.com/learningequality/kolibri/issues/11276
- [x] https://github.com/learningequality/kolibri/issues/11275
- [x] https://github.com/learningequality/kolibri/issues/11274


## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

[Figma
](https://www.figma.com/file/wRlClsHgcVqIwgUrntRN1a/Kolibri-Enhanced-Quiz-Management?type=design&node-id=28-7898&mode=design&t=zsfffy4qmp7fH8xj-0)


## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

- `quizForge` - good or not good name?
- Please test deleting the last item, active items, etc
- Test the side panels and note that the section ID shown should reflect which item is being edited
- Navigation - browser, arrows/X's in side panels, etc

No critical testing needed here yet - this should be merged relatively quickly to unblock follow-up work.

nucleogenesis committed:
remove unnecessary 85px top margin from CoachImmersivePage

nucleogenesis committed:
center and maxat 1000px the kpagecontainer in createquizsection

nucleogenesis committed:
polishing quiz title field & icon

nucleogenesis committed:
tidy spacing on text/title field

nucleogenesis committed:
make the composable useQuizCreation the default export

nucleogenesis committed:
provide/inject composable as `quizForge` - set quiz title on field blur evt

nucleogenesis committed:
remove unused resource selection

further cleaning

futherfurther cleanup

nucleogenesis committed:
KTabs data derived from useQuizCreation; can add sections

nucleogenesis committed:
don't just go back, set the back route correctly

nucleogenesis committed:
tabs have options; addsection focuses new section;

nucleogenesis committed:
show unique value for each section in KTabsPanel to be sure its working

nucleogenesis committed:
don't set added sections as active, let the dev do it manually

nucleogenesis committed:
routing for side panels fixed; hooked into actual UI

nucleogenesis committed:
fix test import

nucleogenesis committed:
moreclean

AllanOXDi commented on kolibri/plugins/coach/assets/src/composables/useQuizCreation.js:
```diff
@@ -124,6 +124,13 @@ export function useQuizCreation() {
       throw new Error(`Section with id ${section_id} not found; cannot be removed.`);
     }
+    if (updatedSections.length === 0) {
+      console.log('Deleting last section...');
```
I think console.log can be removed.

AllanOXDi commented on kolibri/plugins/coach/assets/src/composables/useQuizCreation.js:
```diff
@@ -240,6 +247,10 @@ export function useQuizCreation() {
     get(allSections).find(s => s.section_id === get(_activeSectionId))
   );
+  /** @type {Arrayc<ComputedRef<QuizSection>>} The inactive sections */
+  const inactiveSections = computed(() =>
```
Nice

AllanOXDi commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/SectionSidePanel.vue:
```diff
@@ -31,12 +32,41 @@
+    watch: {
+      $route: function(_, o) {
+        this.prevRoute = o;
+        console.log('prevroute', this.prevRoute);
```
The same here.

AllanOXDi commented on kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/CreateQuizSection.vue:
```diff
@@ -491,19 +296,25 @@
 
   .style-icon {
-    width: 2.5em;
-    height: 2.5em;
```
What could be the reason for not maintaining em units ? as we discussed.

nucleogenesis replied:
Old habits die hard 😅 - will update to use em

thanksameeelian reviewed (commented):
this is all looking pretty good to me! i followed the reviewer guidance and navigated through the available components a few times, successfully added and deleted sections from the beginning, middle, and end, seeing the titles update & the ids be correct, opening the side panel, etc., and i found it was all (or nearly all) working as i expected! 

i'm interested in the outcome of @AllanOXDi's comments, and the ones i'll leave here are all UX-related. i'd like to call out these concerns here to make sure they're visible, but many may be better handled in a separate, follow-up PR (or just dismissed haha)

- i was thinking that if you add a new section, perhaps that section should be automatically focused? but that is just an idea/preference that can be easily rebutted by the design team or by you if you don't feel like that's the behavior you'd like :) 

- i think there is a little bit of focus-ring-funkiness going on, which is probably best displayed in the attached screen recording: 
   - when you tab through the page, the `X` closing button for the whole page is only visibly focused on the first pass through. if you tab back around to it again, it _is_ focused (at the end of the video i trigger the invisibly-focused `X` with my keyboard and successfully close the page), but the only visual indication is the URL preview at the bottom of the window. 
   
   - the focus rings around the section names are currently displaying oddly, and when focus moves to the Edit/Delete menu-triggering button it isn't very clear - it appears that parts/sides of the outline are missing, particularly the side that abuts the next section name. the last section in the list has the most complete outline, but it is still a bit unexpected-looking

https://github.com/learningequality/kolibri/assets/43046460/309f975f-5bfa-4f69-ae85-bade45b6dafa




- just a comment about something that feels weird to me, which is the appearance of hovering over a currently-selected-section's name: 
<img width="191" alt="Screen Shot 2023-09-29 at 12 15 19 PM" src="https://github.com/learningequality/kolibri/assets/43046460/c272eac4-9874-4a66-976e-e195d6508045">


 i just find this visually a little odd. it makes sense that just the name would be showing as hovered/greyed, as the button beside it has a separate hover state which also makes sense, but the disconnect between the underline (of the whole containing div) and what is highlighted in the hover state keeps catching my eye. not sure if there is any possible "better" way to do it though!

nucleogenesis commented:
> i was thinking that if you add a new section, perhaps that section should be automatically focused? but that is just an idea/preference that can be easily rebutted by the design team or by you if you don't feel like that's the behavior you'd like :)

:100: for sure will update accordingly 

To the rest of your comments, which I think are all due to the same weirdness of there being a `<button>` bonanza going on there with one big button holding two little buttons for each tab.

The wrapping <button> is due to how the `KTabs` component works. I'll dive back into this and see if I can work out a way to be sure that the wrapping button is always skipped in the keyboard nav flow. _I think_ it should just never focus the wrapping button because it isn't  handled and for our cases would better be a `<div>`.

I'll also look at the focus ring w/ regard to the page closing X button. Thank you so much for all of your notes!

nucleogenesis committed:
remove console.log, fix jsdoc type

nucleogenesis committed:
remove consolelog

nucleogenesis committed:
px -> em in create quiz section

nucleogenesis committed:
add new section focuses new section

nucleogenesis commented:
@thanksameeelian @AllanOXDi I've addressed most of the feedback here -- the main thing I haven't addressed here is the fix for the tabs. I've begun applying that fix in my ongoing work to close #11274

AllanOXDi reviewed (approved):
The changes have been addressed !

thanksameeelian reviewed (approved):
LGTM!

so much is being introduced here & i'm so excited to see it taking shape. on top of everything else, thanks for your attention to detail & for keeping an eye on that tab issue going forward!
